### PR TITLE
Add `hal::callback_recorder`

### PR DESCRIPTION
**Pull request by MaliaLabor**

### CODE REVIEW

File: include/libhal/callback_recorder.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    return m_callback;
+  }
+
+  /**
+   * @brief Returns true if the callback was ever called
+   *
+   * @return true - was called
+   * @return false - has not been called
+   */
+  [[nodiscard]] bool called_once() noexcept
```
Comment (bryceschober) [inline include/libhal/callback_recorder.hpp]: Just a casual follower chiming in from the "peanut gallery" here, but this seems like a misleading name. In another mocking framework I'm used to, once means exactly once.

Comment (kammce) [reply include/libhal/callback_recorder.hpp]: Hmmm, great point. do you think `was_called()` is a suitable name here? Also might be good to have an `exactly_once()` shorthand function as well.

Comment (bryceschober) [reply include/libhal/callback_recorder.hpp]: Probably `was_called()`. A separate `was_called_once()` and `was_called_n_times(size_t)` have been helpful to me in other contexts.

Comment (bryceschober) [reply include/libhal/callback_recorder.hpp]: ... but I also wouldn't put too much effort into mock sophistication without turning to an established mocking library for help with the grunt work. The one I'm familiar with (CppUTest mock support) wouldn't be appropriate here, but I hear good things about Google Mock and [trompoleil](https://github.com/rollbear/trompeloeil).

**Review by kammce — DISMISSED** [review]

File: include/libhal/callback_recorder.hpp (hunk 1 above)
Comment (kammce) [reply include/libhal/callback_recorder.hpp]: In a previous project, we used FakeIt, but the burden on compile was enormous and it actually made our tests more confusing to work with. So we opted to added in the mocks where we found them necessary. GMock, trompoleil and such are great for interfaces but in this case, we are just checking for functions. We did use [FFF](https://github.com/meekrosoft/fff), but I wanted a C++ version of the library that worked well with lambdas.

If you, I, or anyone else comes across a really nice fake function framework for C++ thats also available on conan, that would be perfect.

**Review by kammce — APPROVED** [review]

**Pull request merged**